[router] Fix byte size logging in HTTP gRPC upstream logs

asraa opened the pull request:
Refreshes byte size on downstream headers in `router.cc`. Enabling gRPC upstream logs will write the byte size of these headers to the logs. Without the refresh, calling `byteSize()` on this header map will result in a bad optional access.

Testing: Added test that configures upstream logs with `envoy.http_grpc_access_log`. Would otherwise result in a crash due to instance of `absl::bad_optional_access`
Fixes
#8828

Signed-off-by: Asra Ali <asraa@google.com>

asraa commented:
CC @ascheglov @mattklein123 
/review @yanavlasov

mattklein123 commented on test/integration/http2_upstream_integration_test.cc:
```diff
@@ -438,4 +438,45 @@ TEST_P(Http2UpstreamIntegrationTest, LargeResponseHeadersRejected) {
+        const std::string json = Json::Factory::loadFromYamlString(yaml_string)->asJsonString();
+        // Replace the terminal envoy.router.
+        hcm.clear_http_filters();
+        TestUtility::loadFromJson(json, *hcm.add_http_filters());
```
nit: you can just use `loadFromYaml` directly

mattklein123 commented on test/integration/http2_upstream_integration_test.cc:
```diff
@@ -438,4 +438,45 @@ TEST_P(Http2UpstreamIntegrationTest, LargeResponseHeadersRejected) {
+  // Send the response headers.
+  upstream_request_->encodeHeaders(default_response_headers_, true);
+  response->waitForEndStream();
+  EXPECT_EQ("200", response->headers().Status()->value().getStringView());
```
In an optimal world we would also actually test the log output, but that's a much larger thing. Maybe add a test TODO and come back to it?

lambdai commented:
/lgtm

asraa commented:
Fixed the test to use YAML directly --
One critical case, Yan/Yuchen maybe can take a deeper look, but by the time `~UpstreamRequest` was called, the object that `parent_.downstream_headers_` points to was deleted...  `async_client_impl_test` failed in CI due to this.
I changed it to a unique pointer and just made an explicit copy inside the filter... PTAL

asraa commented:
@htuch

alyssawilk commented on source/common/router/router.cc:
```diff
@@ -345,7 +345,7 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
   ASSERT(headers.Host());
 
-  downstream_headers_ = &headers;
+  downstream_headers_ = std::make_unique<Http::HeaderMapImpl>(headers);
```
Good catch on use-after deletion!  I'm a bit leery of the copy from a perf perspective.  If Matt thinks this is the right thing to do from an expediency standpoint how about a TODO to see if we can sort out the underlying lifetime issues later?

mattklein123 replied:
We should be logging before teardown. What is the failing call stack?

asraa replied:
[2019-10-31 17:53:38.259][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:83] Caught Segmentation fault, suspect faulting address 0x2337de
[2019-10-31 17:53:38.259][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:70] Backtrace (use tools/stack_decode.py to get line numbers):
[2019-10-31 17:53:38.259][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:71] Envoy version: 0/1.12.0-dev/redacted/RELEASE/BoringSSL
[2019-10-31 17:53:38.259][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #0: __restore_rt [0x7f9428177390]
[2019-10-31 17:53:38.261][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #1: Envoy::Router::Filter::UpstreamRequest::~UpstreamRequest() [0x7cefde]
[2019-10-31 17:53:38.263][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #2: Envoy::Router::Filter::resetAll() [0x7caa9f]
[2019-10-31 17:53:38.264][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #3: Envoy::Router::Filter::onDestroy() [0x7ca9ae]
[2019-10-31 17:53:38.266][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #4: Envoy::Http::AsyncStreamImpl::reset() [0x7c1702]
[2019-10-31 17:53:38.267][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #5: Envoy::Http::AsyncClientImpl::~AsyncClientImpl() [0x7bfb22]
[2019-10-31 17:53:38.269][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #6: Envoy::Http::(anonymous namespace)::AsyncClientImplTest::~AsyncClientImplTest() [0x525b89]
[2019-10-31 17:53:38.270][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #7: Envoy::Http::(anonymous namespace)::AsyncClientImplTest_WatermarkCallbacks_Test::~AsyncClientImplTest_WatermarkCallbacks_Test() [0x570ace]
[2019-10-31 17:53:38.272][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #8: testing::internal::HandleExceptionsInMethodIfSupported<>() [0xc313d4]
[2019-10-31 17:53:38.274][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #9: testing::TestInfo::Run() [0xc32144]
[2019-10-31 17:53:38.275][18][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #10: testing::TestSuite::Run() [0xc32a97]

asraa replied:
I'm honestly inclined to use the manual byte size count in the logger for this headermap and add as a TODO to remove ASAP so at least it doesn't crash in the release.

mattklein123 replied:
From a very quick look this is an order of destuction issue inside async client, and I think the bug is there. I would rather not add the copy to the router.

If there is a quick/slow/localized fix only for the upstream logging, let's just do that w/ a TODO?

asraa replied:
It won't be any slower to call the manual byte size call inside the logger, since the refresh outside the logger does exactly that too.

I'll investigate what the destruction issue is.

mattklein123 replied:
OK yeah we can fix async client later. The issue basically is that in the destroy/reset case we are destroying the headers before the stream in the AsyncRequest case, and possibly not even storing them for the streaming case.

mattklein123 commented:
Replaced by https://github.com/envoyproxy/envoy/pull/8836
